Code/Framework/AzCore/AzCore/Math/Geometry3DUtils.cpp

```diff
@@ -0,0 +1,101 @@
+#include <AzCore/Casting/numeric_cast.h>
+
+namespace AZ::Geometry3DUtils
+{
+    AZStd::vector<AZ::Vector3> GenerateIcoSphere(const uint8_t subdivisionDepth)
+    {
+        // The algorithm for generating an icosphere is to start with a regular icosahedron (20-sided polygon with triangluar faces),
+        // and then for each triangle, subdivide it further into 4 triangles at the midpoint of each edge. This is repeated recursively
+        // to the given depth to reach a desired number of subdivided faces.
+
+        constexpr int IcosahedronFaces = 20;
+
+        // "phi" is also known as the "golden ratio", (1 + sqrt(5)) / 2
+        const float phi = (1.0f + sqrt(5.0f)) * 0.5f;
+
+        // The 12 vertices of an icosahedron centered at the origin with edge length 1 are
+        // (+- 1, +- 1/phi, 0), (0, +- 1, +- 1/phi), (+- 1/phi, 0, +- 1)
+        const AZ::Vector3 icosahedronVertices[] = {
+            // (+- 1, +- 1/phi, 0)
+            { -1.0f, 1.0f/phi, 0.0f }, { 1.0f, 1.0f/phi, 0.0f }, { -1.0f, -1.0f/phi, 0.0f }, { 1.0f, -1.0f/phi, 0.0f },
+            // (0, +- 1, +- 1/phi)
+            { 0.0f, -1.0f, 1.0f/phi }, { 0.0f, 1.0f, 1.0f/phi }, { 0.0f, -1.0f, -1.0f/phi }, { 0.0f, 1.0f, -1.0f/phi },
+            // (+- 1/phi, 0, +- 1)
+            { 1.0f/phi, 0.0f, -1.0f }, { 1.0f/phi, 0.0f, 1.0f }, { -1.0f/phi, 0.0f, -1.0f }, { -1.0f/phi, 0.0f, 1.0f } };
+
+        // The 20 triangles that make up the faces of the icosahedron.
+        int faceIndices[IcosahedronFaces][3] = {
+            // The first 5 faces are around vertex 0
+            { 0, 11, 5 }, { 0, 5, 1 }, { 0, 1, 7 }, { 0, 7, 10 }, { 0, 10, 11 },
+            // The second 5 faces are the ones adjacent to those
+            { 1, 5, 9 }, { 5, 11, 4 }, { 11, 10, 2 }, { 10, 7, 6 }, { 7, 1, 8 },
+            // The third 5 faces are around vertex 3, which is directly opposite to vertex 0
+            { 3, 9, 4 }, { 3, 4, 2 }, { 3, 2, 6 }, { 3, 6, 8 }, { 3, 8, 9 },
+            // The final 5 faces are the ones adjacent to those
+            { 4, 9, 5 }, { 2, 4, 11 }, { 6, 2, 10 }, { 8, 6, 7 }, { 9, 8, 1 }
+        };
+
+        // We're intentionally limiting this algorithm to a max recursion depth of 9 right now as that generates 5M triangles (15M verts)
+        // which should be more than sufficient for most purposes. If it ever turns out not to be enough, this limit can be increased as
+        // needed.
+        uint8_t cappedDepth = AZStd::min(subdivisionDepth, static_cast<uint8_t>(9));
+
+        // Preallocate enough space for all the vertices we're going to generate.
+        AZStd::vector<AZ::Vector3> sphereVertices;
+        sphereVertices.reserve(IcosahedronFaces * 3 * aznumeric_cast<int>(pow(4, cappedDepth)));
+
+        // Recursive lambda method for subdividing each face of the polygon.
+        AZStd::function<void(const AZ::Vector3&, const AZ::Vector3&, const AZ::Vector3&, const uint8_t)> SubdivideTriangle =
```

Comment:
This doesn't have to be stored in `AZStd::function`, this can just be stored in a lambda via `auto`

Reply:
Because this is a recursive lambda, auto doesn't work. It needs an explicit definition.

Reply:
Ok, I understand.
`auto` can't deduce the lambda type until the lambda body has been parsed completely.
But the lambda body can't be parsed completely using `auto` because the call to `SubdivideTriangle` doesn't know the type because it can't be deduced inside of itself.

Reply:
Is there any major cost involved with it being an std function, allocations, capturing, etc? The only variable it seems to be using is sphereVertices, which could just be passed in as a reference if it was a free function.

Reply:
AZStd::function has a 16-byte function buffer for storing function objects that are <= 16 bytes in size as part of the function class.
Because this llambda captures all the local variables, a dynamic allocation does occur, as the sizeof the lambda is greater than 16-bytes(See [functional_base.h](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/function/function_base.h#L201-L210) for more details)

Reply:
Changed to free function. I just had it as a lambda to keep things tidy and self-contained, but I wasn't really invested in the choice. Happier to make it a little more efficient.